### PR TITLE
Sphinx base documentation

### PR DESCRIPTION
**Pull request by Zethson**
Only for discussion for now.

image

This PR adds Sphinx based documentation.
Build inside /docs
```
pip install -r requirements.txt
make html
```

It is hell and unsupported to include markdown files from different directories. Hence, I am using a README.rst file for now. There are some hacks out there on the internet, which use m2r2 or use symlinks to add the content where needed.
But really, it's bad and I could not really get it to work in a reasonable way.
Just like the mkdocs PR I want to have your opinion first and after we decided which way to go I will flesh it out with GH Actions etc.

Now we can compare mkdocs (see PR: https://github.com/ewels/MegaQC/pull/149) and a simple Sphinx based implementation.

I myself love sphinx and use it for all of my own projects, but with reStructuredText. Markdown works, but is clearly 2. class.

Options:

1. We go with mkdocs, which is less powerful than Sphinx, but "just works".
2. We mix rst for the README and markdown for everything else with Sphinx.
3. We go full rst with Sphinx.

Opinions?
Cheers

Signed-off-by: Zethson <lukas.heumos@posteo.net>

### CODE REVIEW

**Comment by apeltzer** [comment]
I think this is mostly for @TMiguelT @tsnowlan and @ewels to decide what they want ;-) 

I'm happy with both solutions, although I also tend to use the tool with the widest support (Sphinx + RST or mkdcos + markdown). 

Benefits of Markdown: 

Markdown is pretty standard on Github, so people could read / correct docs written in Markdown and we could have them directly deployed to a Github page. 

Benefits of RST:

Works nicely with Sphinx ;-) 


My tendency goes with Markdown+mkdocs in this case, as it doesn't look like we really need some of the more advanced features of Sphinx for now. As long as we use standard Markdown, the door to convert at some point (e.g. via `pandoc`) to RST and switch over to Sphinx + RST isn't closed forever. 

Happy to take comments on this, thanks Lukas for making both options available to check / have a look at.

**Comment by tsnowlan** [comment]
I've not worked extensively with either and have no strong (or informed) opinions either way, so I'll defer to Phil and Michael

**Comment by apeltzer** [comment]
Dropping in https://v2.docusaurus.io/docs/ into the ring here, too ;-)

**Comment by Zethson** [comment]
@apeltzer I'll play around with it later :)
Looks pretty powerful.
However, it would not be able to auto generate technical stuff like Sphinx does and @TMiguelT desires.

**Comment by TMiguelT** [comment]
Yep. I would rather we just merged the Sphinx docs so we can move onto the next task rather than trying every documentation builder in existence.

**Comment by Zethson** [comment]
All right. I will polish this PR.
Let's just roll with rst for README + markdown for the rest for now.

Pandoc can 1 to 1 translate everything to rst anyways if we so desire.

**Comment by TMiguelT** [comment]
I'm happy if you convert everything to RST, that doesn't super bother me. Some other features you might look at are:
* The CLI and REST API docs I mentioned earlier
* Auto building and deploying the docs in GitHub Actions

**Comment by Zethson** [comment]
Still a lot to do regarding the documentation, but just wanted to share a couple of successes:

Autogenerated Flask API & Quick API documentation

image

Honestly, it is kind of black magic to me, since I am providing it with a Config class:

```
.. autoflask:: megaqc.app:create_app('megaqc.settings.ProdConfig')
   :undoc-static:
```

Auto generated module documentation with working code links -> if you press source you automatically get a link to the exact definition of the function/class on Github

image

Auto generated Click documentation

image


Yes, the style is TBD. I will most likely revert most of it, since I don't have the time to overwrite most of it.

I will finalize this PR tomorrow. It will get bigger than I expected it to be, but a highly automated solution to last :+1: 

Cheers

**Comment by TMiguelT** [comment]
The autogenerated stuff looks great! I'm no good with colour choice so I wouldn't stray too far from the default theme personally. The default blue/dark grey/white scheme shares some colours with the MegaQC logo so it's not so bad unedited.

**Comment by TMiguelT** [comment]
Oh and also, because the Python API (ie the actual classes and functions) is not a public API, I think it would be good to place it under the title "dev docs", separately from the other public APIs: REST and the CLI. Because it's still useful for developers wanting to read the docstrings etc but it isn't generally useful to users.

File: megaqc/cli.py [hunk 1]
```diff
@@ -28,10 +28,9 @@ def create_megaqc_app(info):
 @click.group(cls=FlaskGroup, create_app=create_megaqc_app)
 def cli():
     """
-    Welcome to the MegaQC command line interface!\n MegaQC is built using the
-    Flask Python web framework.
+    Welcome to the MegaQC command line interface.
 
-    See below for the available commands - for example,
+    \nSee below for the available commands - for example,
```
Comment (Zethson) [inline megaqc/cli.py:33]: docformatter always adds a `.` to the top sentence no matter what.
This looked really ugly in the autogenerated docs then.

Hence, this change.

File: megaqc/rest_api/schemas.py [hunk 2]
```diff
@@ -416,9 +416,9 @@ class FilterObjectSchema(BaseSchema):
     """
     A single filter object, e.g.
     {
-        'type': 'daterange',
-        'value': [],
-        'cmp': 'in'
+    'type': 'daterange',
```
Comment (Zethson) [inline megaqc/rest_api/schemas.py:419]: Sphinx complained when this was intended.
Looks worse now in code, but no warnings when generating the docs. But it's very minor, so whatever.

**Comment by Zethson** [comment]
Ready for review I guess. 
Content wise I am not happy with the documentation, since there is some overlap and the structure is still meh, but this should be addressed in another PR in my opinion.

Please test whether you can build the documentation locally

```
cd docs
pip install -r requirements.txt
make api-docs
make html
```

Documentation should then be in build_/html

Open index.html in your favorite browser.

I did not really test the Github workflow, but it should hopefully work. Would be happy if you could have a look.
Requested @apeltzer since this is not MegaQC code related and I don't want to bother you @TMiguelT too much. Feel free to add comments if you want to.

Cheers.

File: .github/workflows/publish_docs.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+on:
+  push:
+    branches:
+      - master
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    if: "!contains(github.event.head_commit.message, '[skip ci]') && !contains(github.event.head_commit.message, '[ci skip]')"
```
Comment (Zethson) [inline .github/workflows/publish_docs.yaml:11]: Technically this could be added to the other workflows as well.

**Comment by apeltzer** [comment]
>Content wise I am not happy with the documentation since there is some overlap and the structure is still meh, but this should be addressed in another PR in my opinion.

I think having the docs automatically build is one step and as you said, the content is something that will also change, especially if there is now some movement towards rolling out everything in React. Something that could be addressed then largely I suppose :-) 

I'll try to review & test today, probably more towards the evening though.

**Comment by apeltzer** [comment]
Seems to be missing megaqc in my case. Installed a fresh conda environment (pip is a bit difficult here on testing cluster):

```bash
 conda install -c conda-forge sphinx sphinx_rtd_theme sphinx-automodapi sphinxcontrib-napoleon sphinx-click sphinxcontrib-httpdomain
```

That should be fine with the requirements.txt, but is missing `megaqc`:

```bash
make api-docs
sphinx-apidoc -o "source" .. ../*setup* ../*tests*
File source/megaqc.rst already exists, skipping.
File source/megaqc.api.rst already exists, skipping.
File source/megaqc.model.rst already exists, skipping.
File source/megaqc.public.rst already exists, skipping.
File source/megaqc.report_plot.rst already exists, skipping.
File source/megaqc.rest_api.rst already exists, skipping.
File source/megaqc.user.rst already exists, skipping.
File source/megaqc.utils.rst already exists, skipping.
File source/modules.rst already exists, skipping.
(sphinx) [peltzera@bibcinews92001 docs]$ ,ale jt
(sphinx) [peltzera@bibcinews92001 docs]$ make html
Running Sphinx v3.3.1

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/home/peltzera/.conda/envs/sphinx/lib/python3.9/site-packages/sphinx/co
    execfile_(filename, namespace)
  File "/home/peltzera/.conda/envs/sphinx/lib/python3.9/site-packages/sphinx/ut
    exec(code, _globals)
  File "/home/peltzera/MegaQC/docs/source/conf.py", line 4, in <module>
    import megaqc
ModuleNotFoundError: No module named 'megaqc'

make: *** [html] Error 2
```

**Comment by apeltzer** [comment]
Maybe should add `megaqc` in requirements.txt too to make sure this is installed automatically? Or maybe this is because I'm using conda and the recipes there don't define megaqc as dependency.

**Comment by apeltzer** [comment]
Looks as intended & building after the above "fixes" worked fine for me. API docs are there, HTML is also fine with the standard rtd theme, and matches fine with MultiQC. I think this is good & can go in :-)

File: .github/workflows/publish_docs.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.9
+
+      - name: Install pip
+        run: python -m pip install --upgrade pip
+
+      # Required to automatically extract Click CLI and Flask API documentation
+      - name: Install MegaQC
+        run: pip install -e .[dev]
```
Comment (apeltzer) [inline .github/workflows/publish_docs.yaml:27]: Ah this is why it didn't work for me - the GHA code seems to be doing the right thing here and installing dev ;-)

File: docs/source/changelog.rst [hunk 5]
```diff
@@ -0,0 +1,77 @@
+Changelog
+=========
+
+0.4.0
```
Comment (apeltzer) [inline docs/source/changelog.rst:4]: There is not even 0.3 - maybe we can remove this section for now ...? 
```suggestion
0.4.0
```

File: docs/source/changelog.rst [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+Bug Fixes
+~~~~~~~~~
+
+Internal Changes
+~~~~~~~~~~~~~~~~
+
+-  Many and more dependency updates
+
+.. _section-1:
+
```
Comment (apeltzer) [inline docs/source/changelog.rst:24]:
```suggestion

```

File: docs/source/changelog.rst [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+   you can do without a token, but it will be deactivated by default,
+   unless it is the first account created
+
+.. _new-features-1:
+
+New Features
+~~~~~~~~~~~~
+
+-  `[#140]`_ Added a changelog. It’s here! You’re reading it!
+
```
Comment (apeltzer) [inline docs/source/changelog.rst:58]:
```suggestion

-  Sphinx based documentation on Github Pages

```

File: docs/source/changelog.rst [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+-  `[#139]`_ Fixed the user management page (``/users/admin/users``),
+   which lost its JavaScript
+
+.. _internal-changes-1:
+
+Internal Changes
+~~~~~~~~~~~~~~~~
+
+-  Tests for the REST API permissions
+-  Enforce inactive users (by default) in the model layer
```
Comment (apeltzer) [inline docs/source/changelog.rst:73]:
```suggestion
-  Enforce inactive users (by default) in the model layer
-  Many and more dependency updates
```

**Review by apeltzer — APPROVED** [review]
Mostly LGTM, some minor comments (e.g. no changelog for 0.4.X so far required, keep this for 0.3.0 for now I think ...?)

File: .github/workflows/publish_docs.yaml (hunk 4 above)
Comment (Zethson) [reply .github/workflows/publish_docs.yaml:27]: Yeah, forgot to mention that :)

File: docs/source/changelog.rst (hunk 5 above)
Comment (Zethson) [reply docs/source/changelog.rst:4]: There is. Line 25.

**Comment by Zethson** [comment]
@apeltzer the changelog for 0.3 was already there, but I just saw that there was no 0.3 release yet. I will add my stuff to the 0.3 log.

I could add megaqc to the docs requirements, but it makes things much uglier with the workflow, because it would have to upgrade the installed version again with the cloned version to ensure that the API docs are built with the latest code.

I don't expect anybody to build the docs themselves except for developers, who will very likely have a MegaQC installation anyways.

**Comment by apeltzer** [comment]
Thats right and the GHA workflow handles it via `dev` installation - looks good to me then.

**Comment by ewels** [comment]
Looks like I missed the party! Awesome work though @Zethson 🚀 

I love markdown, but I simply don't know rst very well. I have nothing against it - I was suggesting that we stick with markdown as I thought that it would be the path of least resistance. I would prioritise consistency over format so quite happy to switch everything to rst here if others agree.

**Comment by Zethson** [comment]
@ewels it is already pure rst ;)

**Comment by TMiguelT** [comment]
Great work! The docs are a big improvement.

Just some feedback that I'll put here for want of a better place to put it:

* The MegaQC homepage screenshot seems to be broken on the readme page
* Since these docs are now within the installation group, I would rename these pages to remove the "MegaQC Installation: " part 
image
* The web API reference is a good start, but it needs to be namespaced based on the type of API. For instance, the new `/rest_api` and the old `/api` should be under separate subheadings. Actually it's debatable if we even need to document the old API but anyway. It might be possible to do this by passing the blueprint objects individually to the sphinx directive instead of the entire flask app.

**Pull request merged**